---
runtime: shiny
output: html_document
---

# Sequencing of COVID19 genomes in Denmark using the Artic Network protocols
A joint collaboration between Danish Serum Institute, Denish Hospitals and Aalborg University has aimed to sequence all the corona virus genomes from Danish patients during the COVID19 outbreak. The first Danish case was reported February 27 2020. 

```{r message=FALSE,echo=FALSE,warning=FALSE}
library(ggplot2)
library(dplyr)
library(readxl)
library(gridExtra)
library(data.table)
```

```{r message=FALSE,echo=FALSE,warning=FALSE}
# Load general metadata
metadata<-read_excel(path = "metadata.xlsx")
# Load pr base coverage data
covd<-data.frame(scaffold=NULL,position=NULL,coverage=NULL,SampleID=NULL)
for (f in metadata$SampleID) {
  covdtemp<-read.delim(file = paste0("results/coverage/",f,".cov.tsv"),sep = "\t",header = T) %>% mutate(SampleID=f)
  covd<-rbind(covd,covdtemp)
}
genomesize=max(covd$position)
countNs<-data.frame(SampleID=NULL,countN=NULL)
for (f in metadata$SampleID) {
  countNstemp<-read.delim(file = paste0("results/N_counts/",f,"N_count.tsv"),sep = "\t",header = F,col.names = c("fileID","countN")) %>% mutate(SampleID=f) %>% select(SampleID,countN)
  countNs<-rbind(countNs,countNstemp)
}
cov_summary<-covd %>% group_by(SampleID) %>% summarise(data_vol=round(sum(coverage)/10^6),avg_cov=round(sum(coverage)/max(position)),zeros=length(which(coverage==0)))
primerscheme<-read.delim(file = "2020-03-27_nCoV-2019-primer.bed",sep = "\t",header = F,col.names = c("track","start","end","amplicon"))

# Calculate pr amplicon coverage
covd$amplicon<-NA
for (i in 1:dim(primerscheme)[1]) {
  covd$amplicon[covd$position>primerscheme$start[i] & covd$position<primerscheme$end[i]]<-as.character(primerscheme$amplicon[i])
}
ampliconcov<-covd %>% filter(!is.na(amplicon)) %>% group_by(SampleID,amplicon) %>% summarise(coverage=mean(coverage))
thr_cov<-100
n_above_thr<-ampliconcov %>% group_by(SampleID) %>% summarise(n_above_thr=length(which(coverage>thr_cov)))
#
tab_fordisplay<-full_join(metadata,cov_summary) %>% full_join(n_above_thr) %>% full_join(countNs)
names(tab_fordisplay)[which(names(tab_fordisplay)=="n_above_thr")]<-paste0('# amplicons\nabove ',thr_cov,"x\ncoverage")
names(tab_fordisplay)[which(names(tab_fordisplay)=="data_vol")]<-"Data (Mbp)"
names(tab_fordisplay)[which(names(tab_fordisplay)=="avg_cov")]<-"Average\ncoverage"
names(tab_fordisplay)[which(names(tab_fordisplay)=="countN")]<-"Number\nof\nNs"
```

# Sample overview
```{r echo = FALSE}
DT::renderDataTable(tab_fordisplay %>% select(SampleID,Samplename,Sampletype,qPCR_Ct,Primetscheme,`Data (Mbp)`,`Average\ncoverage`,paste0('# amplicons\nabove ',thr_cov,"x\ncoverage"),'Number\nof\nNs'))
```

# Genome coverage
Select the sample for detailed plotting

```{r echo = FALSE}
selectInput("sampleID", label = "Sample ID:",
              choices = metadata$SampleID)
  
```
 
## Per base position 
Coverage plot for each position in the genome
 
```{r echo = FALSE}
renderPlot({
  ggplot(data = covd %>% filter(SampleID==input$sampleID),aes(x = position,y = coverage,col=SampleID))+geom_line()+scale_y_log10()
})
```

## Per amplicon
Average coverage for each 98 amplicons.

```{r fig.width=25,echo = FALSE}
renderPlot({
  ggplot(data = ampliconcov %>% filter(SampleID==input$sampleID),aes(x = amplicon,y = coverage,col=SampleID))+geom_point()+scale_y_log10()+theme(axis.text.x = element_text(size=8,angle = 90))+expand_limits(y=10)+geom_hline(yintercept = 100,linetype="dashed",col="blue")
})
```

# Methods
## Library preparation and sequencing
The RNA was reverse transcribed and PCR amplified according to the Artic Network protocol [https://www.protocols.io/view/ncov-2019-sequencing-protocol-bbmuik6w](https://www.protocols.io/view/ncov-2019-sequencing-protocol-bbmuik6w) with the V2 primer scheme. The DNA libraries were sequenced on the MinION platform Mk1B on R9.4.1 flowcells. The sequencing was carried out with MinKNOW v. .

## Data processing
The data was basecalled using guppy v. 3.4.5 using the high accuracy model. The basecalled data was processed using the "artic-ncov2019-medaka" protocol with the "--minimap2" setting  [https://artic.network/ncov-2019/ncov2019-bioinformatics-sop.html](https://artic.network/ncov-2019/ncov2019-bioinformatics-sop.html). Read coverage was extracted from the bam files using samtools depth with the "-a" option. The processed data was visualised using R v. 3.6.1. The genomes will be submitted to GISAID [https://www.gisaid.org/](https://www.gisaid.org/) and the reads mapping to the virus genome will be submitted to ENA.

