---
title: "QC report - `r params$batch`"
author: "Microflora Corona"
date: "`r format(Sys.time(), '%d-%m-%Y')`, Aalborg, Denmark"
output:
  html_document
params:
  batch: "Batch name"
  labmeta:  "lab metadata"
  input_dir:  "Input directory"
  ref: "genome reference"
editor_options: 
      chunk_output_type: console
---

```{r setup,include = F}
knitr::opts_chunk$set(
  fig.cap    = " ",
  fig.align  = "center",
  fig.width  = 12,
  fig.height = 8,
  message    = FALSE, 
  warning    = FALSE,
  echo       = FALSE)

options(scipen=999)

# Load libraries.
library("tidyverse")
library("magrittr")
library("ggtree")
library("knitr")
library("kableExtra")
library("rjson")
library("ape")
```

```{r,eval=F}
# For build mode.
batch     <- "CJ060"
labmeta   <- "/srv/rbd/covid19/processing/CJ060/sample_sheet.csv"
input_dir <- "/srv/rbd/covid19/processing/CJ060"
ref       <- "/srv/rbd/covid19/git/covid19/workflow/dependencies/ref/MN908947.3.fasta"
```

```{r}
batch     <- params$batch
labmeta   <- params$labmeta
input_dir <- params$input_dir
ref <- params$ref
```

```{r}
# Load amplicon counts.---------------------------------------------------------

cov <- read_delim(file = paste0(input_dir,"/processing/results/amplicon_count.tsv"), delim = "\t") %>%
  mutate(
    library_id = str_match(library_id, "_(.*?)_")[,2],
    primer_id  = str_match(primer_id,"_(.*?)_")[,2] %>% as.numeric()) %>%
  pivot_wider(
    names_from  = "library_id", 
    values_from = "count", 
    values_fill = list("count" = 0)) %>% 
  pivot_longer(
    names_to  = "library_id",
    values_to = "count", 
    cols      = c(-primer_id,-start,-end)) %>%
  dplyr::rename(coverage = count) %>%
  group_by(library_id) %>%
  mutate(
    coverage_relative = coverage/sum(coverage)*100,
    pos_mean          = start + (end-start)/2)

cov_tot <- mutate(cov,
                  bp_temp = (end-start)*coverage %>% round(0)) %>%
           group_by(library_id) %>%
           summarise(coverage = round(sum(bp_temp)/29903,0))
```


```{r}
# Load naive snv frequencies.---------------------------------------------------

nsnv <- read_delim(file = paste0(input_dir,"/processing/results/naive_vcf.tsv"), delim = "\t") %>%
  mutate(library_id = str_match(library_id, "_(.*?)_")[,2])
```


```{r}
# Load artic snv frequencies.---------------------------------------------------

asnv <- read_delim(file = paste0(input_dir,"/processing/results/artic_vcf.tsv"), delim = "\t") %>%
  mutate(
    library_id = str_match(library_id, "_(.*?)_")[,2],
    depth      = str_match(INFO, "DP=(.*?);")[,2] %>% as.numeric(),
    AC         = str_match(INFO, "AC=(.*?);")[,2]) %>%
  separate(AC, into = c("ref_count", "alt_count"), sep = ",", convert = T) %>%
  mutate(alt_frac = round(alt_count/depth, 2)) %>%
  mutate(snv_type = ifelse(nchar(REF) > nchar(ALT), "D", 
                    ifelse(nchar(REF) < nchar(ALT), "I", ALT)),
         het = ifelse(grepl(pattern = "^0/1", x = SAMPLE), "het", "norm")) %>%
  dplyr::rename(position = POS, 
                ref = REF, 
                alt = ALT) %>%
  select(library_id, type, position, ref, alt, depth, ref_count, alt_count, alt_frac, snv_type, het, SAMPLE) %>%
  filter(depth > 10) %>%  
  mutate(position = ifelse(!(snv_type %in% c("D","I")) & nchar(ref) > 1,
                           mapply(function(pos,ref){paste(seq(pos,pos+(nchar(ref)-1)),collapse = ",")},pos = position,ref = ref),position),
         ref      = ifelse(!(snv_type %in% c("D","I")) & nchar(ref) > 1,
                           gsub('(?<=.)(?=.)', ',', ref,perl=TRUE),ref),
         alt      = ifelse(!(snv_type %in% c("D","I")) & nchar(ref) > 1,
                           gsub('(?<=.)(?=.)', ',', alt,perl=TRUE),alt),
        snv_type = ifelse(!(snv_type %in% c("D","I")) & nchar(ref) > 1,
                          alt,snv_type)) %>%
  separate_rows(position,ref,alt,snv_type,sep = ",") %>%
  mutate(position = as.numeric(position))
```


```{r}
# Load N's.---------------------------------------------------------------------
genome_qc <- read_delim(file = paste0(input_dir,"/processing/results/N_counts.tsv"), delim = "\t") %>%
  mutate(library_id = str_match(library_id, "_(.*?)_")[,2]) 
```


```{r}
# Load tree.--------------------------------------------------------------------
ntree <- read.tree(file = paste0(input_dir,"/QC/aligntree/tree_raw.nwk")) 
ntree$tip.label <- ntree$tip.label %>%
  {ifelse(grepl("MN908947",.),.,str_match(.,"_(.*?)_")[,2])}
```


```{r}
# Load the library concentrations.----------------------------------------------
libconc <- read_delim(labmeta, delim = ",", col_names = F) %>%
  {
    if (ncol(.) == 3){
      .
    } else if (ncol(.) == 2) {
      transmute(., X1,X2,X3 = NA)
    }
  } %>%
  `colnames<-`(c("library_id","barcode","library_conc"))
```

```{r}
# Get clade assignment.---------------------------------------------------------
clades <- fromJSON(file = paste0(input_dir,"/QC/aligntree/clades.json")) %>%
  .$nodes %>%
  sapply("[[",1) %>%
  {data.frame(library_id = names(.),Clade = .,row.names = NULL)} %>%
  filter(!grepl("NODE",library_id)) %>%
  mutate(library_id = str_match(library_id, "_(.*?)_")[,2])
```

```{r}
# Get SNV calls from the file.--------------------------------------------------
nt_ref <- ape::read.dna(file = ref, 
                        format = "fasta", as.character = T) %>%
          as.vector() %>% data.frame(ref = .) %>% mutate(position = as.character(1:n()))

snv <- ape::read.dna(file = paste0(input_dir,"/QC/aligntree/aligned.fasta"), 
                     format = "fasta", as.character = T) %>%
          {`colnames<-`(.,as.character(1:ncol(.)))} %>%
          data.frame(check.names = F) %>%
          rownames_to_column("library_id") %>%
          pivot_longer(-library_id, names_to = "position", values_to = "seq") %>%
          left_join(nt_ref, by = "position") %>%
          filter(as.character(ref) != as.character(seq)) %>%
          mutate(seq = toupper(seq),
                 ref = toupper(ref),
                 var = ifelse(seq == "N", "n_count", "snv_count")) %>%
          mutate(position = as.numeric(position)) %>%
          filter(position > 130 & position < 29903-100) %>%
          dplyr::rename(alt = seq) %>%
          mutate(library_id = str_match(library_id, "_(.*?)_")[,2])
```

```{r}
# Merge data.-------------------------------------------------------------------
d <- left_join(libconc, cov_tot, by = "library_id") %>%
     left_join(genome_qc, by = "library_id") %>%
     left_join(clades,by = "library_id") %>%
     mutate(sample_type = tolower(library_id) %>%
       {case_when(grepl("neg",.) ~ "NEG",grepl("pos",.) ~ "POS",!grepl("neg|pos",.) ~ "sample")}) %>%
  mutate(coverage = ifelse(is.na(coverage),0,coverage))

# QC label.
d <- d %>%
  mutate(genome_qc = ifelse(is.na(n_count),
                       ifelse(coverage >= 5 ,"Fail_artic","Fail"),
                       cut(n_count, breaks = c(-Inf,130,5000,Inf), labels = c("HQ","MQ","Fail")) %>% as.character())) %>%
  mutate(genome_qc = factor(genome_qc,levels = c("HQ","MQ","Fail","Fail_artic"))) 
```

# General overview

## Overview of genome qualities

- `HQ`: A genome with < 130 N's in total.
- `MQ`: A genome with < 5000 N's in total.
- `Fail`: Either 1) A genome with > 5000 N's in total or 2) less than 5x average coverage.
- `Fail_artic`: The artic pipeline failed for some reason. 

```{r}
tab <- d %>% 
  group_by(genome_qc,.drop = F) %>%
  summarise(!!batch := n()) 

tot <- tab %>% pull(!!batch) %>% sum() %>%
  data.frame("Total",.) %>% `colnames<-`(colnames(tab)) 

rbind(tab,tot) %>%
  mutate(genome_qc = factor(genome_qc,levels = c("HQ","MQ","Fail","Fail_artic","Total"))) %>%
  arrange(genome_qc) %>%
  dplyr::rename(`Genome Quality` = genome_qc) %>%
  kable() %>%
  kable_styling(bootstrap_options = c("striped", "hover"), full_width = F) %>%
  row_spec(5,bold=T)
```

## Library concentrations vs. coverage

```{r}
ggplot(data = d, aes(x=library_conc, y = coverage, color = genome_qc)) +
  geom_point() +
  labs(y      = "Average coverage",
       x      = "Library Concentration [ng/ul]",
       colour = "Genome Quality") +
  scale_y_log10()
```

## Samples failing the artic pipeline

```{r}
filter(d,genome_qc == "Fail_artic") %>%
  select(library_id,library_conc) %>%
  kable() %>%
  kable_styling(bootstrap_options = c("striped", "hover"), full_width = F)
```

## Coverage for each primer-pair (samples only)

```{r, fig.height = 3}
wh <- filter(d, !is.na(genome_qc) & sample_type == "sample") %>% pull(library_id)

cov_sub <- subset(cov, library_id %in% wh) %>% 
  mutate(coverage_relative = ifelse(coverage_relative < 0.01, 0.01, coverage_relative)) %>%
  left_join(d %>% select(library_id, genome_qc), by = "library_id") %>%
  mutate(genome_qc = factor(genome_qc, levels = c("HQ", "MQ", "Fail")))

ggplot(cov_sub, aes(x = primer_id, y = coverage_relative, color = coverage_relative)) +
  geom_jitter(width = 0.1 , height = 0.1) +
  scale_colour_viridis_b(trans = "log10", direction = -1, breaks = c(0, 0.1, 1, 10, 100)) +
  scale_x_continuous(breaks = c(0,5,10,15,20,25,30)) +
  facet_grid(cols = vars(genome_qc)) +
  scale_y_log10(breaks = c(0.01, 0.1, 1, 10, 100), limits = c(0.005, NA)) +
  labs(x = "Primer pair", y = "Relative coverage [%]") +
  theme(legend.position = "none")
```

## Coverage for each primer-pair (controls only)

```{r, fig.height = 8}
meta_sub <- filter(d,sample_type != "sample")

if (nrow(meta_sub) > 0){ 
  # Subset.
  cov_sub  <- filter(cov, library_id %in% meta_sub$library_id)
  snv_sub <-  filter(snv, library_id %in% meta_sub$library_id)
  
  p <- ggplot(cov_sub, aes(x = pos_mean, y = coverage+1, color = coverage+1)) +
    geom_segment(aes(x = start, xend = end, y = coverage + 1, yend = coverage +1), size = 1) +
    scale_colour_viridis_c(trans = "log10", direction = -1, breaks = c(1, 10, 100, 1000, 10000),limits = c(1, NA)) +
    facet_wrap(~library_id, ncol = 2) +
    scale_y_log10(limits=c(1,100000), breaks = c(1, 10 ,100 ,1000, 10000)) +
    labs(x = "Genome position [bp]", y = "Coverage [fold]") +
    theme(legend.position = "none") +
    geom_text(data = meta_sub, aes(x = 15000, y = 100000, label = paste0("Library conc.: ", round(library_conc, 0), " ng/ul, Coverage: ", coverage)), color = "black", size = 3) 

  if (nrow(snv_sub) > 0){
    p <- p + geom_text(data = snv_sub, aes(x = position, y = 10, label = alt), color = "black", size = 2)
  }
  p
} else{
  "No controls were detected in the data"
} 
```

## Deletions or insertions in the consensus sequence
```{r}
meta_sub <- filter(d,sample_type == "sample" & genome_qc != "Fail")
asnv_sub <- filter(asnv, library_id %in% meta_sub$library_id) %>%
  mutate(tj = paste0(library_id, position))
snv_sub <-  filter(snv, library_id %in% meta_sub$library_id) %>%
  mutate(tj = paste0(library_id, position))

filter(snv_sub, alt == "-" | ref == "-") %>% 
  select(library_id, position, ref, alt)  %>%
  kable() %>%
  kable_styling(bootstrap_options = c("striped", "hover"), full_width = F)
```

## Deletions or insertions marked in the vcf output
```{r}
filter(asnv_sub, snv_type %in% c("I","D")) %>% 
  select(library_id, type, snv_type, position, ref, alt, depth, ref_count, alt_count, alt_frac) %>%
  kable() %>%
  kable_styling(bootstrap_options = c("striped", "hover"), full_width = F)
```

# Detail overview

<!-- ## Clade assignment (MQ and HQ only) -->
<!-- ```{r} -->
<!-- cladedump <- filter(d,sample_type == "sample" & genome_qc %in% c("MQ","HQ")) %>% -->
<!--   select(library_id,Clade) %>%  -->
<!--   dplyr::rename(`Clade assignment` = Clade)  -->

<!-- # Dump .tsv in final output. -->
<!-- write_delim(cladedump,path = paste0(input_dir,"/final_output/clades.tsv"),delim = "\t") -->

<!-- cladedump %>% -->
<!--   kable() %>% -->
<!--   kable_styling(bootstrap_options = c("striped", "hover"), full_width = F) -->
<!-- ``` -->

## Coverage (HQ)

```{r}
meta_sub <- filter(d,sample_type == "sample" & genome_qc == "HQ") %>%
  head(n = 10) 
if (nrow(meta_sub) > 0){
  # Subset.
  cov_sub  <- filter(cov, library_id %in% meta_sub$library_id)
  #asnv_sub <- filter(asnv, library_id %in% meta_sub$library_id)
  snv_sub <-  filter(snv, library_id %in% meta_sub$library_id)

  ggplot(cov_sub, aes(x = pos_mean, y = coverage+1, color = coverage+1)) +
    geom_segment(aes(x = start, xend = end, y = coverage + 1, yend = coverage +1), size = 1) +
    scale_colour_viridis_c(trans = "log10", direction = -1, breaks = c(1, 10, 100, 1000, 10000),limits = c(1, NA)) +
    facet_wrap(~library_id, ncol = 2) +
    scale_y_log10(limits=c(1,100000), breaks = c(1, 10 ,100 ,1000, 10000)) +
    labs(x = "Genome position [bp]", y = "Coverage [fold]") +
    theme(legend.position = "none") +
   geom_text(data = meta_sub, aes(x = 15000, y = 100000, label = paste0("Library conc.: ", round(library_conc, 0), " ng/ul, Coverage: ", coverage)), color = "black", size = 3) +
    #geom_text(data = asnv_sub, aes(x = position, y = 100, label = paste0(snv_type, "\n", alt_frac, "\n", het)), color = "red", size = 2) +
    geom_text(data = snv_sub, aes(x = position, y = 1, label = alt), color = "black", size = 2)
}

```

## Coverage (MQ)

```{r}
meta_sub <- filter(d,sample_type == "sample" & genome_qc == "MQ") %>%
  head(n = 10) 

if (nrow(meta_sub) > 0){
  # Subset.
  cov_sub  <- filter(cov, library_id %in% meta_sub$library_id)
  #asnv_sub <- filter(asnv, library_id %in% meta_sub$library_id)
  snv_sub <-  filter(snv, library_id %in% meta_sub$library_id)

  ggplot(cov_sub, aes(x = pos_mean, y = coverage+1, color = coverage+1)) +
    geom_segment(aes(x = start, xend = end, y = coverage + 1, yend = coverage +1), size = 1) +
    scale_colour_viridis_c(trans = "log10", direction = -1, breaks = c(1, 10, 100, 1000, 10000),limits = c(1, NA)) +
    facet_wrap(~library_id, ncol = 2) +
    scale_y_log10(limits=c(1,100000), breaks = c(1, 10 ,100 ,1000, 10000)) +
    labs(x = "Genome position [bp]", y = "Coverage [fold]") +
    theme(legend.position = "none") +
    geom_text(data = meta_sub, aes(x = 15000, y = 100000, label = paste0("Library conc.: ", round(library_conc, 0), " ng/ul, Coverage: ", coverage)), color = "black", size = 3) +
    #geom_text(data = asnv_sub, aes(x = position, y = 10, label = paste0(snv_type, "\n", alt_frac, "\n", het)), color = "black", size = 2) +
    geom_text(data = snv_sub, aes(x = position, y = 1, label = alt), color = "black", size = 2)
}
```

## Identify mixed-virus samples (SNV artic)

```{r, fig.height = 14}
d_tree <- filter(d, sample_type == "sample")
ntree_trim <- drop.tip(ntree, ntree$tip.label[!(ntree$tip.label %in% d_tree$library_id)]) 

p <- ggtree(ntree_trim) %<+% d_tree

p1 <- p + geom_treescale() +
    geom_tiplab(align = T, linesize = .25, size = 3, aes(color = genome_qc), 
                offset = 0.00003, fontface = "bold", hjust = 1) +
    scale_color_manual(values = c("#2ca25f", "darkred"), name = "Genome Quality") +
    guides(colour = guide_legend(override.aes = list(size =4))) +
    geom_tippoint(aes(color = genome_qc), size = 2) +
    theme(legend.position = c(0.1,0.75))

hq <- filter(d, genome_qc == "HQ")


#Take SNV's seen in at least 2 genomes in the HQ subset ()
snv_filter <- filter(snv, library_id %in% hq$library_id) %>%
              group_by(position) %>%
              summarise(count = n()) %>%
              filter(count >1)

if (nrow(snv_filter) > 0){

snv_heat <- filter(snv, position %in% snv_filter$position & library_id %in% ntree_trim$tip.label) %>%
            select(library_id, position, alt) %>%
            pivot_wider(names_from = "position", values_from = "alt") 

snv_heat2 <- snv_heat %>%
  select(colnames(snv_heat)[-1] %>% as.numeric() %>% sort() %>% as.character())

rownames(snv_heat2) <-snv_heat$library_id

gheatmap(p1, snv_heat2, colnames_angle = 90, hjust = 1, font.size = 2, offset = 0.00004, width = 0.35) +
  scale_fill_manual(values = c("#33a02c", "#1f78b4", "grey30", "#de2d26", "grey80", "orange"),  
                    name = "Variant", breaks = c("A", "C", "G", "T", "N", "-")) +
  theme(legend.position = c(0.1,0.65))
}
```

## Identify mixed-virus samples (SNV freq)

```{r, fig.height = 14}
if (nrow(snv_filter) > 0){
freq_heat <- filter(nsnv, position %in% snv_filter$position & library_id %in% ntree_trim$tip.label) %>%
             select(library_id, position, frac_ALT) %>%
             pivot_wider(names_from = "position", values_from = "frac_ALT") 

freq_heat2 <- freq_heat %>%
  select(colnames(freq_heat)[-1] %>% as.numeric() %>% sort() %>% as.character())

rownames(freq_heat2) <-freq_heat$library_id

p <- ggtree(ntree_trim) %<+% d

p1 <- p + geom_treescale() +
    geom_tiplab(align = T, linesize = .25, size = 3, aes(color = library_conc), 
                offset = 0.00003, fontface = "bold", hjust = 1) +
    geom_tippoint(aes(color = library_conc, shape = genome_qc), size = 2) +
    scale_color_viridis_c(name = "Library concentration [ng/ul] ", direction = -1, 
                          trans = "log10", breaks = c(0.1, 0.3, 1, 2, 5, 15, 50), limits = c(0.1,NA)) +
    scale_shape_discrete(name = "Genome Quality") +
    theme(legend.position = c(0.1,0.75))

gheatmap(p1, freq_heat2, colnames_angle = 90, hjust = 1, font.size = 2, offset = 0.00004, width = 0.35) +
  scale_fill_viridis_c(name = "Alternative variant (SNV)", limits = c(0,1), na.value = "white", direction = -1) +
  theme(legend.position = c(0.1,0.65))
}
```