---
title: "QC report - `r params$batch`"
author: "Microflora Corona"
date: "`r format(Sys.time(), '%d-%m-%Y')`, Aalborg, Denmark"
output:
  html_document
params:
  labmeta:  "lab metadata"
  input_dir:  "Input directory"
  scheme_dir: "Scheme version directory"
  ref: "genome reference"
editor_options: 
      chunk_output_type: console
---

```{r setup,include = F}
knitr::opts_chunk$set(
  fig.cap    = " ",
  fig.align  = "center",
  fig.width  = 12,
  fig.height = 8,
  message    = FALSE, 
  warning    = FALSE,
  echo       = FALSE)

options(scipen=999)

# Load libraries.
library("tidyverse")
library("scales")
library("magrittr")
library("ggtree")
library("knitr")
library("kableExtra")
library("ape")
```

```{r,eval=F}
# For build mode.
input_dir  <- "/srv/rbd/covid19/processing/CJ238"
batch      <- basename(input_dir)
labmeta    <- paste0(input_dir,"/sample_sheet.csv")
scheme_dir <- "/srv/rbd/tym/covid19/workflow/dependencies/primer_schemes/nCoV-2019/aau_long_v3.1"
ref        <- "/srv/rbd/tym/covid19/workflow/dependencies/ref/MN908947.3.fasta"
```

```{r}
input_dir  <- params$input_dir
labmeta    <- params$labmeta
scheme_dir <- params$scheme_dir
ref        <- params$ref

batch <- basename(input_dir)
```

```{r}
# Load amplicon counts.---------------------------------------------------------

cov <- read_delim(file = paste0(input_dir,"/processing/results/amplicon_count.tsv"), delim = "\t") %>%
  mutate(
    library_id = str_match(library_id, "_(.*?)_")[,2],
    primer_id  = str_match(primer_id,"_(.*?)_")[,2] %>% as.numeric()) %>%
  unique() %>% # Need this due to bug that a line sometimes is printed twice.
  complete(library_id,nesting(primer_id,start,end),fill = list(count = 0)) %>%
  dplyr::rename(coverage = count) %>%
  group_by(library_id) %>%
  mutate(
    coverage_relative = coverage/sum(coverage)*100,
    pos_mean          = start + (end-start)/2)
 
# Append the library pool.
libpool <- read_delim(file = paste0(scheme_dir,"/nCoV-2019.scheme.bed"),delim = "\t",col_names = F) %>%
  dplyr::rename(primer_id = X4,pool = X5) %>%
  mutate(primer_id = str_match(primer_id,"_(.*?)_")[,2] %>% as.numeric(),
         pool      = sub("nCoV-2019_","",x = pool) %>% as.numeric()) %>%
  select(primer_id,pool)

cov <- left_join(cov,libpool,by = "primer_id")

tmp1 <- mutate(cov,bp_temp = (end-start)*coverage %>% round(0)) %>%
  group_by(library_id) %>%
  summarise(coverage   = round(sum(bp_temp)/29903,0),
            dropout    = sum(coverage < 20)) 

tmp2 <- mutate(cov,
  bp_temp = (end-start)*coverage %>% round(0),
  pool    = paste0("coverage_",pool)) %>%
  group_by(library_id,pool) %>%
  summarise(coverage   = round(sum(bp_temp)/29903,0),
            dropout    = sum(coverage < 20)) %>%
  pivot_wider(names_from = pool,values_from = c(coverage,dropout)) %>%
  {`colnames<-`(.,sub("_coverage_","_",colnames(.)))}

cov_tot <- left_join(tmp1,tmp2,by = "library_id")
```


```{r}
# Load naive snv frequencies.---------------------------------------------------
nsnv <- read_delim(file = paste0(input_dir,"/processing/results/naive_vcf.tsv"), delim = "\t") %>%
  mutate(library_id = str_match(library_id, "_(.*?)_")[,2])
```


<!-- ```{r} -->
<!-- # Load artic snv frequencies.--------------------------------------------------- -->

<!-- asnv <- read_delim(file = paste0(input_dir,"/processing/results/artic_vcf.tsv"), delim = "\t") %>% -->
<!--   mutate( -->
<!--     library_id = str_match(library_id, "_(.*?)_")[,2], -->
<!--     depth      = str_match(INFO, "DP=(.*?);")[,2] %>% as.numeric(), -->
<!--     AC         = str_match(INFO, "AC=(.*?);")[,2]) %>% -->
<!--   separate(AC, into = c("ref_count", "alt_count"), sep = ",", convert = T) %>% -->
<!--   mutate(alt_frac = round(alt_count/depth, 2)) %>% -->
<!--   mutate(snv_type = ifelse(nchar(REF) > nchar(ALT), "D",  -->
<!--                     ifelse(nchar(REF) < nchar(ALT), "I", ALT)), -->
<!--          het = ifelse(grepl(pattern = "^0/1", x = SAMPLE), "het", "norm")) %>% -->
<!--   dplyr::rename(position = POS,  -->
<!--                 ref = REF,  -->
<!--                 alt = ALT) %>% -->
<!--   select(library_id, type, position, ref, alt, depth, ref_count, alt_count, alt_frac, snv_type, het, SAMPLE) %>% -->
<!--   filter(depth > 10) %>%   -->
<!--   mutate(position = ifelse(!(snv_type %in% c("D","I")) & nchar(ref) > 1, -->
<!--                            mapply(function(pos,ref){paste(seq(pos,pos+(nchar(ref)-1)),collapse = ",")},pos = position,ref = ref),position), -->
<!--          ref      = ifelse(!(snv_type %in% c("D","I")) & nchar(ref) > 1, -->
<!--                            gsub('(?<=.)(?=.)', ',', ref,perl=TRUE),ref), -->
<!--          alt      = ifelse(!(snv_type %in% c("D","I")) & nchar(ref) > 1, -->
<!--                            gsub('(?<=.)(?=.)', ',', alt,perl=TRUE),alt), -->
<!--         snv_type = ifelse(!(snv_type %in% c("D","I")) & nchar(ref) > 1, -->
<!--                           alt,snv_type)) %>% -->
<!--   separate_rows(position,ref,alt,snv_type,sep = ",") %>% -->
<!--   mutate(position = as.numeric(position)) -->
<!-- ``` -->

```{r}
# Load N's.---------------------------------------------------------------------
genome_qc <- read_delim(file = paste0(input_dir,"/processing/results/N_counts.tsv"), delim = "\t") %>%
  mutate(library_id = str_match(library_id, "_(.*?)_")[,2]) 
```


```{r}
# Load tree.--------------------------------------------------------------------
ntree <- read.tree(file = paste0(input_dir,"/QC/aligntree/tree_raw.nwk")) 
ntree$tip.label <- ntree$tip.label %>%
  {ifelse(grepl("MN908947",.),.,str_match(.,"_(.*?)_")[,2])}
```


```{r}
# Load the library concentrations.----------------------------------------------
libconc <- read_delim(labmeta, delim = ",", col_names = F) %>%
  {
    if (ncol(.) == 3){
      .
    } else if (ncol(.) == 2) {
      transmute(., X1,X2,X3 = NA)
    }
  } %>%
  `colnames<-`(c("library_id","barcode","library_conc"))
```

```{r}
# Get SNV calls from the file.--------------------------------------------------
nt_ref <- ape::read.dna(file = ref, 
                        format = "fasta", as.character = T) %>%
          as.vector() %>% data.frame(ref = .) %>% mutate(position = as.character(1:n()))

snv <- ape::read.dna(file = paste0(input_dir,"/QC/aligntree/aligned.fasta"), 
                     format = "fasta", as.character = T) %>%
          {`colnames<-`(.,as.character(1:ncol(.)))} %>%
          data.frame(check.names = F) %>%
          rownames_to_column("library_id") %>%
          pivot_longer(-library_id, names_to = "position", values_to = "seq") %>%
          left_join(nt_ref, by = "position") %>%
          filter(as.character(ref) != as.character(seq)) %>%
          mutate(seq = toupper(seq),
                 ref = toupper(ref),
                 var = ifelse(seq == "N", "n_count", "snv_count")) %>%
          mutate(position = as.numeric(position)) %>%
          filter(position > 130 & position < 29903-100) %>%
          dplyr::rename(alt = seq) %>%
          mutate(library_id = str_match(library_id, "_(.*?)_")[,2])
```

```{r}
# Load nextclade.---------------------------------------------------------------
nextclade <- read_delim(file = paste0(input_dir,"/QC/nextclade.tsv"), delim = "\t") %>%
  dplyr::rename(
    library_id    = seqName,
    pcr_mutations = pcrPrimerChanges) %>%
  mutate(library_id = str_match(library_id, "_(.*?)_")[,2]) 

# primer mutations as data.frame.
primer_muts <- nextclade %>%
  select(library_id,pcr_mutations) %>%
  filter(!is.na(pcr_mutations)) %>%
  mutate(primer_id = lapply(pcr_mutations,function(i){
    strsplit(i,split = ",")[[1]] %>% 
      str_match(pattern = "_(.*?)_") %>% 
      .[,2] %>% 
      data.frame(primer_id = .) %>%
      count(primer_id,name = "n_muts") 
  })) %>%
  unnest(cols = "primer_id") %>%
  mutate(primer_id = as.numeric(as.character(primer_id))) %>%
  select(-pcr_mutations)
```

```{r}
# Merge data.-------------------------------------------------------------------
d <- left_join(libconc, cov_tot, by = "library_id") %>%
     left_join(genome_qc, by = "library_id") %>%
     mutate(sample_type = tolower(library_id) %>%
       {case_when(grepl("neg",.) ~ "NEG",grepl("pos",.) ~ "POS",!grepl("neg|pos",.) ~ "sample")}) %>%
  mutate(coverage   = ifelse(is.na(coverage),0,coverage)) %>%
  mutate(coverage_1 = ifelse(is.na(coverage_1),0,coverage_1)) %>%
  mutate(coverage_2 = ifelse(is.na(coverage_2),0,coverage_2)) %>%
  mutate(dropout    = ifelse(is.na(dropout),unique(libpool$primer_id) %>% length(),dropout)) %>%
  mutate(dropout_1  = ifelse(is.na(dropout_1),unique(libpool$primer_id) %>% length(),dropout_1)) %>%
  mutate(dropout_2  = ifelse(is.na(dropout_2),unique(libpool$primer_id) %>% length(),dropout_2))

  # QC label.
d <- d %>%
  mutate(genome_qc = cut(n_count, breaks = c(-Inf,130,3000,Inf), labels = c("HQ","MQ","Fail")) %>% as.character()) %>%
  mutate(genome_qc = ifelse(is.na(genome_qc),"Fail",genome_qc)) %>%
  mutate(genome_qc = factor(genome_qc,levels = c("HQ","MQ","Fail"))) 

# Dump data.--------------------------------------------------------------------

```

# General overview

## Overview of genome qualities 

- `HQ`: A genome with < 130 N's in total.
- `MQ`: A genome with < 3000 N's in total.
- `Fail`: A genome with >3000 N's in total.

```{r}
# samples.
tab <- d %>% filter(sample_type == "sample") %>%
  group_by(genome_qc,.drop = F) %>%
  summarise(n = n())

tot <- tab %>% pull(n) %>% sum() %>%
  data.frame("Total",.) %>% `colnames<-`(colnames(tab))

tab <- mutate(tab,n = paste0(n," (",round(n/sum(n),1)*100,"%)"))

rbind(tab,tot) %>%
  mutate(genome_qc = factor(as.character(genome_qc),levels = c("HQ","MQ","Fail","Total"))) %>%
  arrange(genome_qc) %>%
  dplyr::rename(`Genome Quality` = genome_qc,!!batch := n) %>%
  kable() %>%
  kable_styling(bootstrap_options = c("striped", "hover"), full_width = F) %>%
  row_spec(4,bold=T,hline_after = T)
```

## Genome quality for controls

```{r}
# Controls.
tab_con <- filter(d,sample_type != "sample")

if (nrow(tab_con) > 0){ 
  tab_con %>% 
    mutate(control = case_when(
      grepl("NEG|neg",library_id) ~ "NEG",
      grepl("POS|pos",library_id) ~ "POS"
    )) %>%
    count(control,genome_qc,.drop = F) %>%
    pivot_wider(names_from = "control",values_from = "n") %>%
    dplyr::rename(`Genome Quality` = genome_qc) %>%
    kable() %>%
    kable_styling(bootstrap_options = c("striped", "hover"), full_width = F) 
} else {
  print("No controls were detected in the data")
}  
```

## Library concentrations vs. coverage

```{r}
cols <- c("#52854C", "#E7B800", "#FC4E07") %>%
  setNames(.,nm = c("HQ","MQ","Fail"))

ggplot(data = d, aes(x=library_conc, y = coverage, color = genome_qc)) +
  geom_point() +
  labs(y      = "Average coverage",
       x      = "Library Concentration [ng/ul]",
       colour = "Genome Quality") +
  scale_y_log10() + 
  scale_colour_manual(values = cols) +
  theme_bw()
```

## Coverage for each primer-pair - controls

```{r, fig.height = 8}
meta_sub <- filter(d,sample_type != "sample")

if (nrow(meta_sub) > 0){ 
  # Subset.
  cov_sub  <- filter(cov, library_id %in% meta_sub$library_id)
  snv_sub <-  filter(snv, library_id %in% meta_sub$library_id)
  
  p <- ggplot(cov_sub, aes(x = pos_mean, y = coverage+1, color = coverage+1)) +
    geom_segment(aes(x = start, xend = end, y = coverage + 1, yend = coverage +1), size = 1) +
    scale_colour_viridis_c(trans = "log10", direction = -1, breaks = c(1, 10, 100, 1000, 10000),limits = c(1, NA)) +
    facet_wrap(~library_id, ncol = 2) +
    scale_y_log10(limits=c(1,100000), breaks = c(1, 10 ,100 ,1000, 10000)) +
    labs(x = "Genome position [bp]", y = "Coverage [fold]") +
    theme(legend.position = "none") +
    geom_text(data = meta_sub, aes(x = 15000, y = 100000, label = paste0("Library conc.: ", round(library_conc, 0), " ng/ul, Coverage: ", coverage)), color = "black", size = 3) 

  if (nrow(snv_sub) > 0){
    p <- p + geom_text(data = snv_sub, aes(x = position, y = 10, label = alt), color = "black", size = 2)
  }
  p
} else{
  print("No controls were detected in the data")
} 
```

## Primer pool ratio (HQ+MQ samples only)

```{r, fig.height=5,fig.width=3}
tmp <- filter(d,genome_qc %in% c("HQ","MQ") & sample_type == "sample") %>% 
  group_by(library_id,genome_qc) %>%
  summarise(coverage_1 = median(coverage_1),coverage_2 = median(coverage_2)) %>%
  mutate(ratio = coverage_1/coverage_2)

tmp_long <- tmp %>%
  pivot_longer(cols = starts_with("coverage"),names_to = "Pool") %>%
  mutate(Pool = gsub("coverage_","",Pool))

ggplot(tmp,aes(x = "",y = ratio)) +
  geom_boxplot(outlier.shape = NA) +
  geom_point(position = position_jitter(width = .1)) +
  geom_hline(yintercept = 1,linetype = "dotted",colour = "red") +
  labs(x = "",y = "Pool 1/2 ratio") +
  scale_y_log10() +
  theme_bw()
```

## Primer pool coverage (HQ+MQ samples only)

```{r, fig.height=5,fig.width=4}
ggplot(tmp_long,aes(x = Pool,y = value)) +
  geom_boxplot(outlier.shape = NA) +
  geom_point(position = position_jitter(width = .1)) +
  scale_y_log10() +
  labs(y = "Average coverage") +
  theme_bw()
```

## Coverage for each primer-pair (HQ+MQ samples only)

```{r}
wh <- filter(d, genome_qc %in% c("HQ","MQ") & sample_type == "sample") %>% pull(library_id)
cov_sub <- subset(cov, library_id %in% wh) %>% 
  mutate(coverage_relative = ifelse(coverage_relative < 0.01, 0.01, coverage_relative)) %>%
  left_join(d %>% select(library_id, genome_qc), by = "library_id") %>%
  mutate(genome_qc = factor(genome_qc, levels = c("HQ", "MQ", "Fail")))

muts_sub <- filter(primer_muts,library_id %in% wh)

covmuts <- left_join(cov_sub,muts_sub,by = c("library_id","primer_id")) %>%
  mutate(n_muts = ifelse(is.na(n_muts),0,n_muts)) %>%
  mutate(n_muts = cut(n_muts,breaks = c(-Inf,0.5,1.5,3.5,Inf),labels = c("0","1","2-3",">3"))) %>%
  arrange(n_muts)

ggplot(covmuts, aes(x = primer_id, y = coverage_relative, color = n_muts)) +
  geom_jitter(width = 0.1 , height = 0.1) +
  scale_colour_manual(values = c("gray","#52854C", "#E7B800", "#FC4E07"),drop = F) +
  scale_x_continuous(breaks = c(0,5,10,15,20,25,30)) +
  #facet_grid(rows = vars(genome_qc)) +
  scale_y_log10(breaks = c(0.01, 0.1, 1, 10, 100), limits = c(0.005, NA)) +
  labs(x = "Primer pair", y = "Relative coverage [%]",colour = "Number of mutations in primer-pair") +
  theme_classic() +
  theme(legend.position = "bottom") 
  
```

## Identify mixed-virus samples (SNV artic)

```{r}
# Pre-define fig height from data.
d_tree <- filter(d, sample_type == "sample" & genome_qc != "Fail")
ntree_trim <- drop.tip(ntree, ntree$tip.label[!(ntree$tip.label %in% d_tree$library_id)]) 

figh <- length(ntree_trim$tip.label)*0.2
```

```{r, fig.height = figh}
p <- ggtree(ntree_trim) %<+% d_tree

p1 <- p + geom_treescale() +
    geom_tiplab(align = T, linesize = .25, size = 3, aes(color = genome_qc), 
                offset = 0.00016, fontface = "bold", hjust = 1) +
    scale_color_manual(values = c("#2ca25f", "darkred"), name = "Genome Quality") +
    guides(colour = guide_legend(override.aes = list(size = 4))) +
    geom_tippoint(aes(color = genome_qc), size = 2) +
    theme(legend.position = c(0.1,0.75))

hq <- filter(d, genome_qc == "HQ")

#Take SNV's seen in at least 2 genomes in the HQ subset ()
snv_filter <- filter(snv, library_id %in% hq$library_id) %>%
              group_by(position) %>%
              summarise(count = n()) %>%
              filter(count >1)

if (nrow(snv_filter) > 0){

snv_heat <- filter(snv, position %in% snv_filter$position & library_id %in% ntree_trim$tip.label) %>%
            select(library_id, position, alt) %>%
            pivot_wider(names_from = "position", values_from = "alt") 

snv_heat2 <- snv_heat %>%
  select(colnames(snv_heat)[-1] %>% as.numeric() %>% sort() %>% as.character())

rownames(snv_heat2) <-snv_heat$library_id

gheatmap(p1, snv_heat2, colnames_angle = 90, hjust = 1, font.size = 2, offset = 0.0002, width = 0.8) +
  scale_fill_manual(values = c("#33a02c", "#1f78b4", "grey30", "#de2d26", "grey80", "orange"),  
                    name = "Variant", breaks = c("A", "C", "G", "T", "N", "-")) +
  theme(legend.position = c(0.1,0.65))
}
```

## Identify mixed-virus samples (SNV freq)

```{r, fig.height = figh}
if (nrow(snv_filter) > 0){
freq_heat <- filter(nsnv, position %in% snv_filter$position & library_id %in% ntree_trim$tip.label) %>%
             select(library_id, position, frac_ALT) %>%
             pivot_wider(names_from = "position", values_from = "frac_ALT") 

freq_heat2 <- freq_heat %>%
  select(colnames(freq_heat)[-1] %>% as.numeric() %>% sort() %>% as.character())

rownames(freq_heat2) <-freq_heat$library_id

p <- ggtree(ntree_trim) %<+% d

p1 <- p + geom_treescale() +
    geom_tiplab(align = T, linesize = .25, size = 3, aes(color = library_conc), 
                offset = 0.00016, fontface = "bold", hjust = 1) +
    geom_tippoint(aes(color = library_conc, shape = genome_qc), size = 2) +
    scale_color_viridis_c(name = "Library concentration [ng/ul] ", direction = -1, 
                          trans = "log10", breaks = c(0.1, 0.3, 1, 2, 5, 15, 50), limits = c(0.1,NA)) +
    scale_shape_discrete(name = "Genome Quality") +
    theme(legend.position = c(0.1,0.75))

gheatmap(p1, freq_heat2, colnames_angle = 90, hjust = 1, font.size = 2, offset = 0.0002, width = 0.8) +
  scale_fill_viridis_c(name = "Alternative variant (SNV)", limits = c(0,1), na.value = "white", direction = -1) +
  theme(legend.position = c(0.1,0.65))
}
```