---
output:
  word_document:
    reference_docx: ssi_report_word_template.docx
  html_document:
    df_print: paged
always_allow_html: yes
editor_options: 
  chunk_output_type: console
knit: (function(inputFile, encoding) { rmarkdown::render(inputFile, encoding = encoding, output_file = file.path(dirname(inputFile), paste0(format(Sys.Date(),"%d-%m-%Y"),"_ugerapport.docx"))) })
---

```{r,include = F}
### USER SPECIFIED PARAMETERS ##################################################

# Set week to focus at, according to the ISO standard. Can be multiple.
weeks <- c(47)

# !OBS: The linelist, genome metadata and phylogenetic tree are automatically loaded from the latest data build.
```

```{r libraries, warning = F, message=F,include=F}
library(tidyverse)
library(readxl)
library(data.table)
library(lubridate)
library(jsonlite)
library(httr)
library(stringr)
library(Biostrings)
library(ggmap)
library(ggtree)
library(ape)
library(kableExtra)
library(ISOweek)

knitr::opts_chunk$set(
  fig.cap    = " ",
  fig.align  = "center",
  #fig.width  = 12,
  #fig.height = 8,
  message    = FALSE, 
  warning    = FALSE,
  echo       = F)

# Format weeks.
date_rng <- paste0("2020-W",c(head(weeks,n=1),tail(weeks,n=1)),c("-1","-7")) %>%
  ISOweek2date() %>%
  as.Date()

prev_week     <- min(weeks) - 1
date_rng_prev <- paste0("2020-W",c(head(prev_week,n=1),tail(prev_week,n=1)),c("-1","-7")) %>%
  ISOweek2date() %>%
  as.Date()
```

```{r read_data}
# Read data
source("read_data.R")
```

```{r, eval = T}
## Temp clade hack
meta_ll$clade <- sub("\\.EU[1-2]", "", meta_ll$clade )
```


```{r read_old_variant, eval = T}
# Read variant definitions
variant_files <- list.files("/srv/rbd/covid19/metadata/cluster_definition",pattern = "_definition", full.names = T)
name_variants_old <- read_csv(variant_files[length(variant_files)]) %>% 
  select(substitutions, variant_no)
```


```{r gen_new_variant}
# Summarize variants and merge old definitions
name_variants <- meta_ll %>%
  filter(!is.na(substitutions)) %>%
  group_by(substitutions) %>%
  summarise(variant_aa_mutation = dplyr::first(aa_mutations),
            variant_size = n(), 
            variant_start = min(as.Date(date_linelist), na.rm = T),
            variant_end = max(as.Date(date_linelist), na.rm = T)) %>%
  arrange(variant_start) %>% 
  left_join(name_variants_old, by = "substitutions")

# Add numbers for new variants
name_variants$variant_no[is.na(name_variants$variant_no)] <- 
  (1:length(name_variants$variant_no[is.na(name_variants$variant_no)]) + max(name_variants$variant_no, na.rm = T))
  
# Add variant name
name_variants <- name_variants %>% 
  mutate(variant_name = paste0("variant ",variant_no)) 
# Dump variant names
write_csv(name_variants, path = file.path("/srv/rbd/covid19/metadata/cluster_definition",
                                          paste0(format(Sys.Date(),"%Y-%m-%d"),"_variant_definition.csv")))
```


```{r merge_variant}
meta_ll <- left_join(meta_ll,name_variants,by = "substitutions") %>% 
  group_by(firstDayWeek) %>%
  mutate(variant_cov     = sum(!is.na(variant_name))/n()) %>%
  ungroup() %>% group_by(firstDayWeek,Region) %>%
  mutate(variant_cov_reg = sum(!is.na(variant_name))/n()) %>%
  ungroup() %>% 
  filter(date_linelist <= date_rng[2])
```

# Genomisk overblik for SARS-CoV-2 i Danmark
Danish Covid-19 Genome Consortium

Rapport udfærdiget af AAU og SSI d. `r format(Sys.time(), '%d-%m-%Y')`

## Baggrund
For at assistere epidemiologer og sundhedsmyndigheder med at overvåge spredningen af SARS-CoV-2 i Danmark blev Danish Covid-19 Genome Consortium (DGGC) etableret som et multidisciplinært samarbejde mellem eksperter fra Statens Serum Institut og Aalborg Universitet. Formålet med samarbejdet er at sekventere SARS-CoV-2 genomer for at se hvordan de udvikler sig, og kombinere denne information med epidemiologiske data, for at kunne følge lokale og globale smittekæder. I indeværende rapport gives et overblik for mængden og kvaliteten af det indsamlede materiale for de fem danske regioner, samt den regionale udvikling af klader. Da der løbende sekventeres prøver kan tallene fra uge til uge ændre sig også bagudrettet.

## Uge statistik
```{r basicStats}
nPos <- nrow(meta_ll)

nSeq <- meta_ll %>% 
  pull(library_id) %>% unique() %>%  length()

nGenomes <- meta_ll %>% 
  filter(is.na(sequence_exclude) & genome_qc %in% c("MQ", "HQ")) %>%
  nrow()

nGenomesThisWeek <- meta_ll %>%
  filter(is.na(sequence_exclude) & genome_qc %in% c("MQ", "HQ")) %>%
  filter(date_linelist >= date_rng[1]) %>%
  nrow()

nDirMut  <- meta_ll %>% filter(!is.na(substitutions) & 
                                  duplicated(substitutions) == F) %>% nrow()

nDirMutOld <- meta_ll %>% filter(date_linelist < date_rng[1] & 
                                    !is.na(substitutions) 
                                  & duplicated(substitutions) == F) %>% nrow()
```

I denne rapport indgår data fra perioden `r format(min(meta_ll$date_linelist), '%d-%m-%Y')` til `r format(date_rng[2], '%d-%m-%Y')`. Der er særligt fokus på uge `r weeks` (`r format(date_rng[1], '%d-%m-%Y')` til `r format(date_rng[2], '%d-%m-%Y')`). Data der anvendes i denne rapport inkluderer:

* Metadata fra linelisten, senest opdateret d. `r format(as.Date(timestmp_ll), '%d-%m-%Y')`
* Fylogenetisk træ, senest opdateret d. `r format(as.Date(timestmp_nxt), '%d-%m-%Y')`

Data for perioden indeholder `r nPos` positive prøver, hvoraf der er sekventeret virus fra `r nSeq`. Dette gav anledning til `r nGenomes` genomer, hvoraf `r nGenomesThisWeek` er tilføjet i uge `r weeks`. Samtlige genomer fordeler sig på `r nDirMut` unikke variantkombinationer mod `r nDirMutOld` i ugen før.

\newpage

# Coronavirus typer over tid og regioner
I figurerne nedenfor vises antallet af de 5 overordnede klader defineret af nextstrain grupperet per måned og region.

```{r summarize_clades, fig.height = 8, dpi=300}
cumclade <- filter(meta_ll,!is.na(firstDayWeek) & !is.na(Region)) %>% 
  filter(!is.na(clade)) %>% 
  select(clade,firstDayWeek, Region) %>%
  mutate(clade = substr(clade,1,3),
         n = 1) %>%
  group_by(Region, firstDayWeek,clade) %>%
  summarise(n = sum(n)) %>%
  ungroup() %>%
  complete(Region, firstDayWeek,clade,fill = list(n = 0)) %>%
  group_by(Region, clade) %>%
  arrange(as.numeric(firstDayWeek)) %>%
  mutate(cs = cumsum(n)) %>%
  group_by(Region,firstDayWeek) %>%
  mutate(freq = n/sum(n),cfreq = cs/sum(cs))

cumclade_DK <- group_by(cumclade,firstDayWeek,clade) %>%
  summarise(n = sum(n)) %>%
  group_by(clade) %>%
  mutate(cs = cumsum(n)) %>%
  group_by(firstDayWeek) %>%
  mutate(freq = n/sum(n),cfreq = cs/sum(cs),
         Region = "Hele Danmark")

cumclade_DK <- bind_rows(cumclade_DK, cumclade)
```

```{r plot_clades, fig.height = 8, fig.width=7, dpi=300}
p <- ggplot(cumclade_DK,aes(x = firstDayWeek, y = n, group = clade, fill = clade)) +
    geom_area() + 
  facet_wrap(~Region, ncol = 2, scales = "free_y") + 
    theme_classic() +
    scale_x_date(date_breaks = "1 month",date_labels = "%b") +
    scale_fill_manual(values = unname(branded_colors), name = "Klade") +
    theme(legend.position = "bottom") +
    ylab("Antal Genomer") +
    xlab("Måned")

plot(p)
```

# Hurtigt voksende smittekæder
Figuren nedenfor viser udviklingen af varianter over tid. Rækker angiver varianter (ens genomer) sorteret indenfor de overordnede virusklader, mens søjlerne angiver fortløbende uger siden starten af epidemien. De enkelte varianter er arbitrært navngivne, og størrelsen på punktet angiver hvor mange sekventerede prøver der var for den givne variant/uge, mens farven angiver hvor stor en procentdel af det samlede antal prøver i en uge en variant udgør.

```{r filter_variants_time}
## Identify clades with more than 10 samples in the period.
clade_freq <- meta_ll %>% 
  filter(date_linelist >= date_rng[1]  & date_linelist <= date_rng[2] & is.na(sequence_exclude)) %>% 
  group_by(variant_no,variant_start,variant_end,clade) %>% 
  summarize(n = n()) %>% 
  #filter(n >= 10 & !is.na(variant_no)) %>%
  arrange(-n)

#ungroup(clade_freq) %>%
#  select(variant_no,clade,n,variant_start) %>%
#  kable(col.names = c("variant", "Klade", "antal","Først observeret")) %>% 
#    kable_paper("hover", full_width = T)
```

```{r plot_variants_time, fig.height=10, fig.width = 12, dpi=300}
# Make a nation-wide figure, all variants >=20.
d_clades_week <- filter(meta_ll, !is.na(clade) & date_linelist <= date_rng[2]) %>%
  filter(!is.na(variant_no)) %>%
  group_by(firstDayWeek,clade,variant_no,variant_size,variant_start) %>%
  summarise(count = n()) %>%
  group_by(firstDayWeek) %>%
  mutate(rel = count/sum(count)*100) %>%
  group_by(variant_no) %>%
  mutate(mutsum = sum(count)) %>%
  #filter(mutsum >= 20) %>%
  filter(mutsum >= 20 | variant_no %in% clade_freq$variant_no[1:5]) %>%
  mutate(txt = paste0("variant ",as.character(variant_no), ", n = ", as.character(variant_size))) %>%
  ungroup() %>%
  arrange(variant_no,clade) %>%
  mutate(txt = factor(txt,levels = unique(txt))) 

ggplot(
  data = d_clades_week,
    aes(
      x     = substr(date2ISOweek(firstDayWeek),1,8) %>% sub(pattern = "W", replacement = ""),
      y     = txt,
      color = rel,
      size  = count)) +
    geom_point() +
    scale_y_discrete(position = "right") +
    scale_size_area(breaks = c(1, 10, 50), max_size = 10) +
    #scale_x_date(date_breaks = "1 week",date_labels = "%W",expand = c(0.01,0.01)) +
    scale_color_viridis_c(trans = "log10", breaks = c(1, 10, 50),) +
    labs(
      title = "Der vises kun varianter med mere end 20 personer samt top5 for den aktive uge",
      color = "Procendel af sekventerede prøver pr. uge",
      x     = "Uge", 
      size  = "Antal") +
    facet_grid(rows = vars(clade),scales = "free_y",space = "free_y") +
    theme(text = element_text(size = 10),
          axis.title.y = element_blank(),
          axis.text.y = element_text(hjust = 0, size = 6),
          axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1),
          legend.position = "top")
```



```{r tab_variants}
# Per region.
meta_table <- left_join(clade_freq,meta_ll, by = "variant_no") %>% 
  filter(date_linelist >= date_rng[1]  & date_linelist <= date_rng[2] & !is.na(Region) & !is.na(variant_no)) %>% 
  group_by(Region,variant_no) %>% 
  summarize(n = n(),
            pct_M = sum(Sex == "M") / n,
            pct_DK = sum(CitizenshipText == "Danmark") / n,
            alder  = median_iqr(as.numeric(SampleAge)))

# Hele DK.
heleDK <- left_join(clade_freq,meta_ll, by = "variant_no" )%>% 
  filter(date_linelist >= date_rng[1]  & date_linelist <= date_rng[2] & !is.na(variant_no)) %>% 
  group_by(variant_no) %>% 
  summarize(n = n(),
            pct_M = sum(Sex == "M") / n,
            pct_DK = sum(CitizenshipText == "Danmark") / n,
            alder  = median_iqr(as.numeric(SampleAge))) %>%
  arrange(-n)

#tables <- list()
#
#tables[["Hele Danmark"]] <- heleDK %>%
#  kable(digits = 2, col.names = c("variant", "n", "Mænd (%)", "Dansk (%)", "Alder (median, iqr)")) %>% 
#  kable_paper("hover", full_width = T)
#
#for (reg in unique(meta_table$Region)){
#  tables[[reg]] <- filter(meta_table,Region == reg) %>%
#    ungroup() %>% select(-Region) %>% arrange(-n) %>%
#    kable(digits = 2, col.names = c("variant", "n", "Mænd (%)", "Dansk (%)", "Alder (median, iqr)")) %>% 
#    kable_paper("hover", full_width = T)
#}
```
\newpage

## Geografisk spredning af varianter
```{r}
nClust <- 5
```

Figuren nedenfor viser den geografiske spredning for de `r nClust` hyppigste varianter i uge `r weeks`.

```{r plot_clusers_map, fig.height=7, fig.width = 7, dpi=300}
meta_map <- meta_ll %>%
  mutate(variant_no = gsub("variant ","",x = variant_name)) %>%
  filter(date_linelist >= date_rng[1]  & date_linelist <= date_rng[2] &
           variant_no %in% heleDK$variant_no[1:nClust]) %>%
  group_by(variant_name) %>%
  mutate(n = n()) %>%
  ungroup() %>%
  arrange(-n) %>%
  mutate(variant_no = paste0(variant_no," (n = ",n,")")) %>%
  mutate(variant_no = factor(variant_no,levels = unique(variant_no)))

p <- ggplot(data = meta_map, aes(x= zip_lon, y = zip_lat, color = variant_no)) +
  #p <- ggplot(data = meta_map, aes(x= lon_adr, y = lat_adr, color = variant_no)) +
  geom_polygon(data = dk_nuts2, inherit.aes = F,
               aes(x = long, y = lat, group = group), color = "black", fill = "grey95") +
  geom_jitter(size = 2,height = 0.05,width = 0.05) +
  #scale_color_viridis_d() +
  scale_color_manual(values = unname(branded_colors)) + 
  labs(colour = "Variant") +
  theme_void() +
  theme(legend.position = c(1,1),legend.justification = c(1.5,1.5))

if (nrow(heleDK) == 0){
  qplot(1,1,shape = NA) + 
    annotate("text",label = "No variants detected in the selected period.",x = 1,y = 1) + 
    theme_void()
} else {
  p
} 
```

